### PR TITLE
Add P4_16 + v1model arch test program for testing multiple "packet pa…

### PR DESCRIPTION
**Pull request by jafingerhut**
…ths"

i.e. resubmit, recirculate, ingress-to-egress clone, egress-to-egress
clone, and multicast replication.  It has a P4_16 program plus an STF
test that checks that output packets are as they should be, checked
pretty carefully by hand.

### CODE REVIEW

**Comment by mbudiu-vmw** [comment]
The STF files were first defined informally and the bmv2stf program was created. Then a more formal implementation was made in stf_parser.py, but the bmv2stf program was never converted to use the new implementation, and now it only handles a subset.  It is actually pretty easy to use, but we always had more pressing problems. If you want to extend the STF grammar we should probably rewrite bmv2stf as well, it's too difficult to keep two versions in sync.

File: testdata/p4_16_samples/v1model-special-ops-bmv2.p4 [hunk 1]
```diff
@@ -0,0 +1,470 @@
...
+            standard_metadata.deq_qdepth : exact;
+            standard_metadata.ingress_global_timestamp : exact;
+            standard_metadata.egress_global_timestamp : exact;
+            standard_metadata.lf_field_list : exact;
+            standard_metadata.mcast_grp : exact;
+            standard_metadata.resubmit_flag : exact;
+            standard_metadata.egress_rid : exact;
+            standard_metadata.checksum_error : exact;
+            standard_metadata.recirculate_flag : exact;
+            //standard_metadata.parser_error : exact;
```
Comment (mbudiu-vmw) [inline testdata/p4_16_samples/v1model-special-ops-bmv2.p4]: Filed #1490 to keep track of this.

File: testdata/p4_16_samples/v1model-special-ops-bmv2.p4 [hunk 2]
```diff
@@ -0,0 +1,470 @@
...
+
+
+control fill_ipv4_address(out bit<32> ipv4_address,
+                          in bit<8> byte0,    // most significant byte
+                          in bit<8> byte1,
+                          in bit<8> byte2,
+                          in bit<8> byte3)    // least significant byte
+{
+    apply {
+        ipv4_address = (((bit<32>) byte0 << 24) |
```
Comment (mbudiu-vmw) [inline testdata/p4_16_samples/v1model-special-ops-bmv2.p4]: must easier to use concatenation `++`

File: testdata/p4_16_samples/v1model-special-ops-bmv2.p4 [hunk 3]
```diff
@@ -0,0 +1,470 @@
...
+        // BMV2_V1MODEL_INSTANCE_TYPE_RESUBMIT.  The instance_type
+        // field is an exception to the "preserve metadata field
+        // value" rule.
+
+        // For the resubmit operation, standard_metadata.resubmit_flag
+        // is 0 for the resubmitted packet, which avoids that packet
+        // being resubmitted indefinitely, unless your program
+        // explicitly causes a separate call to resubmit() on each
+        // execution of ingress.
+        resubmit(standard_metadata);
```
Comment (mbudiu-vmw) [inline testdata/p4_16_samples/v1model-special-ops-bmv2.p4]: You should an explicit type argument to resubmit if you want field names.
`resubmit<standard_metadata_t>(standard_metadata)`

File: backends/bmv2/bmv2stf.py [hunk 4]
```diff
@@ -387,6 +387,14 @@ def do_command(self, cmd):
             self.do_cli_command(self.parse_table_add(cmd))
         elif first == "setdefault":
             self.do_cli_command(self.parse_table_set_default(cmd))
+        elif first == "mirroring_add":
```
Comment (mbudiu-vmw) [inline backends/bmv2/bmv2stf.py]: Filed #1489 to unify this script with the stf parsing library

**Comment by mbudiu-vmw** [comment]
I see no reason not to merge this.

File: testdata/p4_16_samples/v1model-special-ops-bmv2.p4 (hunk 3 above)
Comment (jafingerhut) [reply testdata/p4_16_samples/v1model-special-ops-bmv2.p4]: I tried making that change to the source code, and at least with the version of p4c I compiled it with, it made no difference in the resulting compiled BMv2 JSON file.  It contained the list of field names with or without the `<standard_metadata_t>`.

Comment (jafingerhut) [reply testdata/p4_16_samples/v1model-special-ops-bmv2.p4]: I believe at least one of the intermediate P4_16 programs generated by the compiler (maybe after front end, maybe after mid end) has this `<standard_metadata_t>` type specifier added to the P4_16 code automatically, so it shouldn't depend upon the back end.

File: testdata/p4_16_samples/v1model-special-ops-bmv2.p4 (hunk 2 above)
Comment (jafingerhut) [reply testdata/p4_16_samples/v1model-special-ops-bmv2.p4]: I am making that change and updating this PR soon.  Doh!  Can't believe I forgot about concatenation while writing out that expression.

File: testdata/p4_16_samples/v1model-special-ops-bmv2.p4 (hunk 3 above)
Comment (mbudiu-vmw) [reply testdata/p4_16_samples/v1model-special-ops-bmv2.p4]: that makes a difference when you use a constant (list) as an argument.

**Pull request merged**